modules/core/src/main/scala/org/scalasteward/core/data/Version.scala

```diff
@@ -157,61 +149,34 @@ object Version {
     final case class Separator(c: Char, startIndex: Int) extends Component
     case object Empty extends Component { override def startIndex: Int = -1 }
 
-    def parse(str: String): List[Component] = {
-      @tailrec
-      def loop(
-          rest: List[Char],
-          accN: List[Char],
-          accA: List[Char],
-          acc: List[Component],
-          index: Int
-      ): List[Component] =
-        rest match {
-          case h :: t =>
-            h match {
-              case '.' | '-' | '_' | '+' =>
-                val (newAcc, newIndex) = materialize(accN, accA, acc, index)
-                loop(t, List.empty, List.empty, Separator(h, newIndex) :: newAcc, newIndex + 1)
-
-              case _ if h.isDigit && accA.nonEmpty =>
-                val (newAcc, newIndex) = materialize(List.empty, accA, acc, index)
-                loop(t, h :: accN, List.empty, newAcc, newIndex)
-
-              case _ if h.isDigit && accA.isEmpty =>
-                loop(t, h :: accN, List.empty, acc, index)
-
-              case _ if accN.nonEmpty =>
-                val (newAcc, newIndex) = materialize(accN, List.empty, acc, index)
-                loop(t, List.empty, h :: accA, newAcc, newIndex)
-
-              case _ if accN.isEmpty =>
-                loop(t, List.empty, h :: accA, acc, index)
-            }
-          case Nil =>
-            val (newAcc, _) = materialize(accN, accA, acc, index)
-            newAcc
-        }
-
-      def materialize(
-          accN: List[Char],
-          accA: List[Char],
-          acc: List[Component],
-          index: Int
-      ): (List[Component], Int) =
-        if (accN.nonEmpty) (Numeric(accN.reverse.mkString, index) :: acc, index + accN.length)
-        else if (accA.nonEmpty) (Alpha(accA.reverse.mkString, index) :: acc, index + accA.length)
-        else (acc, index)
-
-      loop(str.toList, List.empty, List.empty, List.empty, 0).reverse
-    }
+    private val numeric = """^(\d+)(.*)$""".r
+    private val separator = """^([.\-_+])(.*)$""".r
+    private val alpha = """^([^.\-_+\d]+)(.*)$""".r
+    private val hash = """([-+])(g?\p{XDigit}{6,})(.*)$""".r
+
+    def parse(str: String, index: Int = 0): List[Component] =
+      str match {
+        case "" => List.empty
+        case hash(sep, value, rest) if !startsWithDate(value) =>
+          Separator(sep.head, index) +: Hash(value, index + 1) +:
+            parse(rest, index + value.length + 1)
+        case numeric(value, rest) =>
+          Numeric(value, index) +: parse(rest, index + value.length)
+        case alpha(value, rest) =>
+          Alpha(value, index) +: parse(rest, index + value.length)
+        case separator(value, rest) =>
+          Separator(value.head, index) +: parse(rest, index + value.length)
+      }
 
     def render(components: List[Component]): String =
       components.map {
         case n: Numeric   => n.value
         case a: Alpha     => a.value
+        case h: Hash      => h.value
```

Comment:
This should be covered by by the tests if #1804 is merged and `master` is merged into this branch.